Add OVL library

chances opened the pull request:
See #3.

## Port OVL Dumper

  - [x] Open OVL files
  - [x] #5
  - [x] #6

## Read RCT3's distributed OVLs

  - [x] `Main.*.ovl`
  - [x] `nullbmp.*.ovl`
  - etc.

chances committed:
Add .editorconfig

chances committed:
Port OVL file reading from rct3-importer

https://github.com/chances/rct3-importer

chances committed:
Update docs

chances committed:
Merge branch 'main' into ovl/trunk

chances committed:
Add known game data paths

chances committed:
Update docs

chances committed:
Create ovl.yml

chances committed:
Merge branch 'ovl/trunk' into ovl/ci

chances committed:
Add OVL dumper utility project

chances committed:
Merge pull request #1 from open-rct3/ovl/ci

Add CI workflow for libOVL

chances committed:
Refactor OVL dumper app

chances committed:
Refactor fields to keep things DRY

chances committed:
Make OVL path parsing more robust

chances committed:
Add more known RCT3 installation paths

chances committed:
Search for RCT3 assets given known paths

chances committed:
Fix how OVL files are opened

jonwil commented:
I would suggest that the code here https://sourceforge.net/p/rct3/code/HEAD/tree/trunk/RCT3%20Importer/src/libOVLng/ (and maybe other code in that repository, don't know) would be a better place to start as it seems to be a lot more comprehensive than what you have here.

GGuyX commented:
Could this potentially let us also finally open up original RCT3 .ovl files after all those years (fuck Frontier for not wanting to allow anyone to mess with their files YEARS after the game released) or would it just be it's own version of the format specifically for open-source version of RCT3, incompatible with original files? I don't have high hopes for this letting us crack open the original game but it's still nice to dream.

jonwil commented:
People have been messing with the original game OVL files ever since I wrote the first mod tools for this game back in the day.

GGuyX commented:
> People have been messing with the original game OVL files ever since I wrote the first mod tools for this game back in the day.

Wasn't aware of that, but thanks for letting me know.

chances commented:
I've hit a speed bump related to the branching logic dependent on OVL archive versions.

This feature is on a brief hiatus while I prototype the game's desktop client.

chances committed:
Render a splash screen with lit

https://lit.dev

chances committed:
Integrate lit SSR

chances committed:
Add a debug task

chances committed:
Simplify and refactor splash screen

chances committed:
Extract an abstract window interface

chances committed:
Add reference screenshots

chances committed:
Port website to Lume

https://lume.land

chances committed:
Finish navigation header reveal animation

chances committed:
Fix header navigation alignment

chances committed:
Fix jumbotron position on medium screens

chances committed:
Fix hamburger icon position when nav is shown

chances committed:
Merge branch 'website/game/trunk' into gui/park/trunk

chances committed:
Remove Diet views in lieu of static Lume site

chances committed:
Compile website with server compilation

chances committed:
Load environment variables from `.env`, if found

chances committed:
Flash errors to the session

chances committed:
Fix `running` flag state in event loop

chances committed:
Add website and GUI mockup readmes

chances committed:
Enhance header reveal interaction

Introduces a focus-based interaction to reveal the site's navigation.

Mouse interactions would sometimes conflict with the new focus-based interaction, i.e. clicking the hamburger nub wouldn't reveal the nav on the first click.

chances committed:
Add an icon

chances committed:
Fix conflict between page title and alerts' titles

chances committed:
Set a troubleshoot action when WebGPU is unavailable

chances committed:
Fix anchor, button, and input focus & outline styles

chances committed:
Add HUD to game GUI

Modified from the mockup GUI.

chances committed:
Enhance site menu title

chances committed:
Fix cursor on unchecked icon switches

chances committed:
Hide GUI panels, by default

chances committed:
Update notes

chances committed:
Watch for website changes in dev mode

chances committed:
Refactor: Move web client

chances committed:
Refactor: Extract major colors

chances committed:
Add an accessible "Skip to content" link

chances committed:
Integrate WMR dev server

chances committed:
Update dev instructions and add a Roadmap outline

chances committed:
Fix interrupt signal handling, Pipe WMR output

Killing the app would leave orphaned `deno` processes. These changes make sure we gracefully end the file watcher and front-end module bundler.

This also reroutes WMR output through the main process.

chances committed:
Add WebStorm project configs

chances committed:
Fix default formatter

chances committed:
Refactor desktop subproject

chances committed:
Add Roadmap commentary

chances committed:
Add inspiration photos

chances committed:
Unify Deno configs into project-level config

chances committed:
Add Deno to IDE environment

chances committed:
Update WebStorm config

chances committed:
Scaffold a .NET conversion

chances committed:
Add an web view to the inspector sidebar

chances committed:
Add an inspector interface model

chances committed:
Add OVL primitives

chances committed:
Add OVL dumper app

chances committed:
Ensure project window is the app's main window

chances committed:
Refactor dumper app, target macOS 10.15, etc.

chances committed:
Update notes and TODOs

chances committed:
Fix OVL header parsing

chances committed:
Read `style.common.ovl` archive

chances committed:
Detect OVL types

chances committed:
Refactor OVL file format handling

chances committed:
Create icons for the app, Dumper, and the Importer

chances committed:
Modernize OVL file type specifier

chances committed:
Use hashing to compare OVL documents

chances committed:
Add error handling extension methods

chances committed:
Refactor OVL fields

chances committed:
Rename document window controller

chances committed:
Remove ineffectual document-handling code

chances committed:
Fix OVL document display name property

chances committed:
Turn on tracing messages when debugging

chances committed:
Refactor error-related extensions

chances committed:
Fix file opening logic

chances committed:
Implement `applicationShouldHandleReopen`

chances committed:
Refactor static getters

chances committed:
Remove debug logging

chances committed:
Display OVL name in its document window

chances committed:
Add OVL file type icons

chances committed:
Add a Project outline view

chances committed:
Make project collection serializable

chances committed:
Add some more file menu actions

chances committed:
Add a Project document type and refactor documents

chances committed:
Use pascal case for actions

chances committed:
Refactor app reactivity

chances committed:
Upgrade dependencies

chances committed:
Simplify target frameworks

chances committed:
Terminate app when last window is closed

chances committed:
Add an OVL Dumper client for Windows

chances committed:
Refactor project structure

chances committed:
Add dumper icon

chances committed:
Add menu strip and status bar

Try to read the OVL archive selected by the user

chances committed:
Use the single-threaded apartment mode for the WinForms client

chances committed:
Make file collection public

chances committed:
Add Windows resources to Dumper

chances committed:
Add a WinForms client

chances committed:
Set startup object and Windows target platform version

chances committed:
Convert from WebGPU to OpenGL

chances committed:
Extract OpenGL functionality into a GLSurface user control

chances committed:
Update TODOs

chances committed:
Merge branch 'dotnet/opentk-gl-control' into ovl/dumper

chances committed:
Show OVL files grouped by file tag

chances committed:
Add all known file types

chances committed:
Ignore test coverage artifacts

chances committed:
Add AI-assisted porting effort summary

chances committed:
Enhance dumper app icons

chances committed:
Exit the dumper app appropriately

chances committed:
Upgrade cache actions

chances committed:
Use .NET 8 to run tests

- Only build on macOS and Windows
- Upload code coverage to [Codecov](https://about.codecov.io)

chances committed:
Allow tests to target either macOS or Windows

chances committed:
Try to load actual resource names from OVLs

chances committed:
Fix macOS compilation

chances committed:
Improve the UX for OVL pairs

- Surface the document name in window titles
- Common and unique resources are split in the tree view
- Added icons to resource types
- Added tooltips with resource details

chances committed:
Improve UX of automatic tree view sizing

chances committed:
Add resource counts to the Dumper status bar

chances committed:
Fix the unit tests

chances committed:
Restructure the testing infrastructure

chances committed:
Improve the test bench UX

chances committed:
Improve the test bench UX

chances committed:
Upgrade dependencies

chances committed:
Refactor test bench form

chances committed:
Add AI docs and plans

chances committed:
Refactor the OVL test bench

chances committed:
Architect an OVL file viewer plugin system

chances committed:
Remove the macOS client

chances committed:
Merge branch 'ovl/dumper' into ovl/dumper-viewer-plugins

# Conflicts:
#	Dumper/Dumper.csproj

chances committed:
Use the web view to render empty content prompts

chances committed:
Add constraints to the sidebar splitter

chances committed:
Fix layout of file content controls

chances committed:
Adjust empty view styles

chances committed:
Pluralize file types in for empty viewers

chances committed:
Prompt the user to open an OVL when none are open

chances committed:
Scaffold a couple OVL viewer plugins using AssemblyScript

chances committed:
Fix plugin compilation on macOS

chances committed:
Fix and parallelize plugin compilation

chances committed:
Update TODOs

chances committed:
Address a few TODOs

- Flesh out the inspector view-model
- Improve icon pixel formats, and
- Reference the OVL library in the OpenRCT3 client project

chances committed:
Add a readme for Dumper plugins

chances committed:
Implement manifold mesh, sound, and spline viewers

chances committed:
Remove redundant TODOs

chances committed:
Ignore plugin artifacts

chances committed:
Bundle the app into the bin folder

chances committed:
Add a plugin integration test

chances committed:
Make the integration test runner cross-platform

chances committed:
Fix debug and test Makefile tasks and update TODOs

chances committed:
Try to switch to Silk.NET

chances committed:
Update dependency locks

chances commented:
I'm back from my hiatus and have made good progress on this PR.

---

Does anyone have any more feedback before I merge the new work and close this issue?

chances committed:
Finish porting Windows client to OpenGL

chances committed:
Fix and improve plugin integration tests

chances committed:
Update TODOs

chances committed:
Merge remote-tracking branch 'origin/ovl/phase-one/trunk' into ovl/phase-one/switch-to-silk

# Conflicts:
#	OpenRCT3/OpenRCT3.csproj
#	OpenRCT3/Platforms/Windows/GLSurface.cs

chances committed:
Delete completed plans

chances committed:
Delete completed plans

chances committed:
Remove dead code

chances committed:
Remove dead code

chances committed:
Fix macOS compilation

chances committed:
Draft an attribution document

chances committed:
Update TODOs

chances committed:
Update readme

chances committed:
Fix solution compilation

chances committed:
Remove proprietary data

chances committed:
Update TODOs and AI rules

chances committed:
Remove invalid testing project

chances committed:
Remove invalid testing project

chances committed:
Simplify TODO.md

chances committed:
Refactor project names and remove dead code

chances committed:
Add Nova tasks

chances committed:
Try to fix test bench CLI

chances committed:
Fix plugin type checking via `deno check`

chances committed:
Refactor to simplify build

chances committed:
Merge branch 'ovl/phase-one/trunk' into ovl/phase-one/switch-to-silk

chances committed:
Write a plan to render an empty flat park

chances committed:
Update AI plans

chances committed:
Remove revision history

That's what Git is for...

chances committed:
Cross-link between plan and cost docs

chances committed:
Scaffold a new GDK project

See https://github.com/open-rct3/open-rct3/blob/ovl/phase-one/trunk/.opencode/plans/render-flat-empty-park.md

chances committed:
Update plans and docs

chances committed:
Use generic filename

chances committed:
Implement Phase 3: Detect RCT3 Installation

chances committed:
Refactor solution structure

chances committed:
Amend empty park plan

- Fix weird class names
- Clarify requirements
- Enhance readability

chances committed:
Use "OpenGL" casing for type names

chances committed:
Add `OSX` and `MACOS` constants

So that we can do platform switching inline:
```cs
#if WINDOWS
// ...
#elif OSX
// ...
#endif

chances committed:
Integrate config loading and RCT3 install search

chances committed:
Integrate rolling logs with NLog

chances committed:
Integrate folder picker fallback into app startup

This ensures users can manually locate their RCT3 installation before the main game logic begins.

chances committed:
Add dead-simple game world primitives

chances committed:
Remove dead code

chances committed:
Implement Phase 4: Render a solid color plane

chances committed:
Fix windows startup behavior

chances committed:
💅 Lint sources

chances committed:
Refactor how the game's renderer is setup

chances committed:
Integrate a weak renderer ref into `Game`, ...

Make `AppConfig` a singleton

chances committed:
Make unowned resources explicit

chances committed:
Set proper window clipping styles for OpenGL rendering

chances committed:
Add a game loop

chances committed:
Reduce CPU usage by sleeping when ahead of schedule

chances committed:
Catch all uncaught exceptions

chances committed:
Add context requested and rendered events

chances committed:
Pump WinForms events

chances committed:
Adjust CPU usage mitigation logic

chances committed:
Fix macOS compilation

chances committed:
Implement a first pass at parsing `nullbmp` OVLs

chances committed:
Add extracted PNGs

chances committed:
Reorganize plans

chances committed:
Duplicate color palette conversion in plugin JS

To reduce the maintenance burden of compiling dot net code for plug insurance, it's easier just to duplicate the singular function.

chances committed:
💅 Format sources

chances committed:
Add a plugins dialog with a fancy new path truncation label

chances committed:
Add plugin toolbar

- Add an `Enabled` property to `IViewerPlugin` and `ViewerPlugin`
- Enhance `PluginsDialog` with a toolbar containing install,
toggle-active, and uninstall actions
- Move plugin loading to `MainForm_Load` with a retry/cancel error dialog
- Use `PuzzleOutline` icon variant for the dialog

chances committed:
Build plugins before build

chances committed:
Fix label truncation logic

chances committed:
Refactor how icons are generated

chances committed:
Replace enabled toggle button with checkbox

Replace `toggleActive` toolbar button with an `enabled` checkbox in
the plugin metadata panel. The checkbox's tooltip is based
on the plugin's current state.

Also:
- Extract `UpdateEmptyState()` from `PluginListBox_SelectedIndexChanged`
- Align Uninstall button to the right of the tool strip

chances committed:
Prevent overflow on the main toolbar

chances committed:
Set icons for plugin installation, Refactor icon renderer

chances committed:
Fix empty state, show "No plugins" appropriately

chances committed:
Prevent toolbar tool tips from obstructing things

chances committed:
Use color icon for "Install from catalog"

chances committed:
Fix plugin artifact nesting in solution explorer

chances committed:
Refactor sources to reduce redundancies

chances committed:
Fix plugin compilation with `env::abort` host function

chances committed:
Improve plugin initialization and error handling

chances committed:
Fix OVL file viewer plugins' metadata functions

chances committed:
Fix OVL resource extraction, Integrate it into Dumper

chances committed:
Add flexi-texture decoder

chances committed:
Fix palette decoding: 1024-byte RGBA instead of 768-byte RGB

Palette format was incorrectly defined as 256x3 (RGB) but FlexiTexture.cs uses 256x4 (RGBA). Also handle case where `hasPalette` is false by outputting black pixels.

Fixes: ftx-viewer: render() nullbmp test

chances committed:
Add a Dumper debugging profile

chances committed:
Unify NuGet packages and remove System.Drawing.Common deps

chances committed:
Save previously selected files for better repeat UX

chances committed:
Reduce flickering when initializing web view

chances committed:
Add a janky wav audio player to the sound plugin

chances committed:
Show only plugin name in content header

chances committed:
Improve fuel error presentation

chances committed:
Refactor OVL decoder

chances committed:
Use template strings instead of string concatenation

chances committed:
Add a shortcut to test OVL parsing

chances committed:
Try to simplify OVL archive parser

chances committed:
Try to thoroughly document the OVL file format

chances committed:
Fact-check archive format

Fixed significant symbol table and relocation errors

chances committed:
🎉 Fix the broken OVL parser

chances committed:
Add `Ovl.ReadResource` method, Refactor OVL parser

chances committed:
Integrate OVL API changes across the solution

chances committed:
Merge branch into ovl/trunk

Merge C# conversion.

# Merge branch 'ovl/phase-one/files/plugin-enhancements' into ovl/trunk
#
# Conflicts:
#	.gitignore
#	dub.json
#	dub.selections.json
#	lib/ovl/.gitignore
#	lib/ovl/dumper/dub.json
#	lib/ovl/dumper/src/app.d
#	lib/ovl/src/ovl/enums.d
#	lib/ovl/src/ovl/gui.d
#	lib/ovl/src/ovl/package.d
#	src/paths.d

chances committed:
Refactor Ovl API, implement the dictionary interface

chances committed:
Refactor away switches in favor of pattern matching

chances committed:
Generate a file name for OVL resources

chances committed:
Populate last opened OVL file name

chances committed:
Normalize viewer UA stylesheet

chances committed:
Merge branch 'ovl/phase-one/fix/ftx-bitmaps' into ovl/trunk

chances committed:
Fix unit tests

chances committed:
Only run Dumper UI tests on windows

chances committed:
Upgrade to the .NET 10 SDK

chances committed:
Skip tests with explicit warnings if RCT3 is not found

chances committed:
Upgrade .NET and setup Deno in CI

chances committed:
Cache Deno dependencies

chances committed:
Rename continuous integration workflow, Upgrade test runners

chances committed:
Ignore slnx.sqlite
